content/blog/2019/nov/grpc-a-powerful-way-to-improve-golang-apis/index.md

```diff
@@ -0,0 +1,286 @@
+
+The strict use of HTTP 2.0 enables developers to embrace some of its features, mainly streaming (client-side and server-side). For example:
+
+- A client can send a bulk of messages through a stream, instead of sending one request each or one with all of them
+- A client can send a request and receive a stream of responses, instead of waiting for the server to finish the processing and send a huge payload
+
+Protocol Buffers are a simpler and optimised way to define and serialise structured data. It can be used not only with GRPC but for other use cases, such as event modelling and data storage.
+
+Think Swagger (REST) or WSDL (SOAP), but lightweight due to the fact the payloads are binary, allowing faster serialisation and small footprint. Being generated by tools such as `protoc`, it brings some benefits such as type-safety in compile-time and auto-completion in IDEs.
+
+In the snippet below, a GRPC service is defined using Protobuf. Generating a client and server requires a simple `protoc` call. Easy right?
+
+```protobuf
+syntax = "proto3";
+
+package api;
+
+// Defines where your go package will be placed after compiled
+option go_package = "api";
+
+service Identity {
+  rpc GetUser (GetUserRequest) returns (GetUserResponse) {}
+  rpc GetUsers (GetUsersRequest) returns (GetUsersResponse) {}
+}
+
+message User {
+  string user_id = 1;
+  string name = 2;
+  bool active = 3;
+}
+
+message GetUserRequest {
+  string user_id = 1;
+}
+
+message GetUserResponse {
+  User user = 1;
+}
+
+message GetUsersRequest {}
+
+message GetUsersResponse {
+  // Defines an array of users
+  repeated User users = 1;
+}
+```
+
+In any protobuf file, each field has a number associated with it. These are unique identifiers in the binarised message and should not be changed. If one of these gets deprecated, they can't be re-used. It helps on avoiding breaking changes, allowing a progressive model evolution. Developers don't need to worry about fields been re-used, clashing with previous payloads, for example.
+
+# Creating a simple GRPC server and client in Golang
+
+_[Code of this part available at Github brunoluiz/grpc-example@master](https://github.com/brunoluiz/grpc-example)_
+
+It is possible to create a simple GRPC server and client based on the previous protobuf definition. The following tools need to be installed:
+
+- [Protocol buffers compiler][1]
+- [Golang protoc plugin][2]
+
+In a Golang project, running `protoc -I. --go_out=plugins=grpc:./generated api/api.proto` will generate both server and client. The `go_package` option in the protobuf specifies the generated code output path. In this case it will be `./generated/api`.
+
+To create a server, all methods defined by the proto need to implemented in a Golang struct. Peaking into the generated files, this is the interface generated from `service Identity`. An implementation fulfilling these methods will be enough to implement a server.
```

Comment:
need to be implemented by a go type... A type that implements this interface is all you need to create to implement the server.